final-project/application.py

```diff
@@ -34,19 +34,94 @@ def after_request(response):
 # configure CS50 Library to use SQLite database
 db = SQL("sqlite:///final.db")
 
+# users table
+# CREATE TABLE users (id INTEGER PRIMARY KEY NOT NULL, username TEXT UNIQUE NOT NULL, email TEXT UNIQUE NOT NULL, hash TEXT NOT NULL, first TEXT, last TEXT);
+
 @app.route("/")
 @login_required
 def index():
 
     # keep user logged in
     session.get("user_id")
 
-    """Render map."""
-    if not os.environ.get("API_KEY"):
-        raise RuntimeError("API_KEY not set")
-    return render_template("index.html", key=os.environ.get("API_KEY"))
+    return apology("TODO")
+
+
+@app.route("/categories")
+@login_required
+def categories():
+
+    # keep user logged in
+    session.get("user_id")
+
+    return apology("TODO")
+
+@app.route("/product")
+# @login_required
+def product():
+
+    # keep user logged in
+    session.get("user_id")
+
+    # get name of product
+    product = "%" + request.args.get("product") + "%"
+
+    try:
+        #Get product from the database
+        product_info = db.execute("SELECT * FROM products WHERE product_name LIKE :product", product = product)
+        # Check whether products were found
+        product_check = product_info[0]
+
+    except RuntimeError:
+        # If problem with db.execute
+        return apology("Error: We'll fix this. Please try again shortly.")
+
+    except IndexError:
+        # If no products were found
+        return apology("No products named " + request.args.get("product"))
```

Comment:
Woah. This throws if there are no products? That's not a good practice. Instread, you should just use an if statement to check here.

Reply:
haha yeah, I thought it was a little hacky when I wrote it. I'll improve the code.

Reply:
🏁 done